### PR TITLE
upgrading to java 11. DBZ-1402

### PR DESCRIPTION
**Pull request by brbrown25**
@jpechane This is an initial commit upgrading to java 11. `mvn clean install -Pjdk11` is being used to  test this out. One thing that is puzzling me is that `debezium-embedded` fails to build this way. However if I go directly into that directory and execute the clean install, it works without error. 

For now I made it so you can build across multiple java versions but can remove all of that and just make it java 11 at the end of reviewing this.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -51,8 +52,8 @@
     <properties>
         <!-- Instruct the build to use only UTF-8 encoding for source code -->
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <maven.compiler.source>1.8</maven.compiler.source>
-        <maven.compiler.target>1.8</maven.compiler.target>
+        <maven.compiler.source>11</maven.compiler.source>
```
Comment (gunnarmorling) [inline pom.xml]: I don't think we should require Java 11 at *runtime* just yet, as it quite drastically reduce the number of potential users (many are still on Java 8 for whatever reasons). At this point, I'd prefer to make sure that Debezium can be *built* with JDK 11 (and here I'm fine with requiring JDK 11 as a baseline, as the number of contributors is much smaller than the number of users).

File: pom.xml [hunk 2]
```diff
@@ -436,6 +454,13 @@
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-surefire-plugin</artifactId>
                     <version>${version.surefire.plugin}</version>
+                    <dependencies>
+                        <dependency>
+                            <groupId>org.ow2.asm</groupId>
+                            <artifactId>asm</artifactId>
+                            <version>6.2</version>
```
Comment (gunnarmorling) [inline pom.xml]: Can that version be extracted into a property, as it's used multiple times. And out of curiousity, why is it needed at all?

File: pom.xml [hunk 3]
```diff
@@ -607,6 +637,10 @@
                     <!--runOrder>alphabetical</runOrder-->
                     <useFile>false</useFile>
                     <enableAssertions>true</enableAssertions>
+                    <!-- https://stackoverflow.com/questions/53437819/maven-surefire-and-jdk-11-->
+                    <!-- https://winterbe.com/posts/2018/08/29/migrate-maven-projects-to-java-11-jigsaw/-->
+                    <forkCount>0</forkCount> <!-- changed this to 0 -->
```
Comment (gunnarmorling) [inline pom.xml]: I like the comments above about the reasoning, but "changed this to 0" is merely stating the obvious ;)

File: pom.xml [hunk 4]
```diff
@@ -755,13 +789,35 @@
...
             <activation>
                 <jdk>9</jdk>
             </activation>
             <properties>
                 <modules.argline>--add-modules java.xml.bind</modules.argline>
+                <maven.compiler.source>9</maven.compiler.source>
+                <maven.compiler.target>9</maven.compiler.target>
+            </properties>
+        </profile>
+        <profile>
```
Comment (gunnarmorling) [inline pom.xml]: Do we actually need all these profiles? How about the suggestion above: keep JDK 8 compatibility at runtime, but require JDK 11 or later at build time (can be required via the Maven enforcer plug-in)?

File: pom.xml (hunk 1 above)
Comment (brbrown25) [reply pom.xml]: Yeah that’s a good point! I’ll update that!

- [ ] Target java 8 for runtime still

File: pom.xml (hunk 2 above)
Comment (brbrown25) [reply pom.xml]: @gunnarmorling i think it’s safe to remove. I had seen some places online where that was added. I had hoped it might help with the debezium embedded build issues. I’m gonna pull it out i think. 

- [ ] verify if necessary. If not remove, if it is extract version.

File: pom.xml (hunk 3 above)
Comment (brbrown25) [reply pom.xml]: Haha...good point there!

File: pom.xml (hunk 4 above)
Comment (brbrown25) [reply pom.xml]: Nope. I think I can just update the jdk9 profile to be 11 and then the Travis build can use that Profile for builds. Then everyone else can continue with java 8 for now. 

- [ ] Clean up profiles.

**Comment by brbrown25** [comment]
thanks @gunnarmorling ill get on these changes this evening and also make the branch to address the debezium container upgrades!

**Comment by brbrown25** [comment]
- [ ] figure out java11 resource bug

**Comment by brbrown25** [comment]
Ok it looks like when compiling the embedded module from the parent it's not running `[DEBUG] adding entry simple/test/a/connector.properties` etc, which it does if running from that module directory, so investigating it a bit further!

**Comment by brbrown25** [comment]
I've figured it out. it appears that the path resolution is acting bizarre in jdk11 namely when invoking the module tests from the parent it properly resolves in jdk8 i.e `src/test/resources/simple/test/a` -> `debezium/debezium-embedded/src/test/resources/simple/test/a` but in jdk11 we get `src/test/resources/simple/test/a` -> `debezium/src/test/resources/simple/test/a`

**Comment by brbrown25** [comment]
@gunnarmorling I'll investigate this pathing issue more this evening, but the latest code should build 🤞

**Comment by gunnarmorling** [comment]
Hey @brbrown25, how are things? Seems this last test failure might be unrelated? Do you observe it locally too on your branch?

File: debezium-embedded/src/test/java/io/debezium/embedded/ConnectorOutputTest.java [hunk 5]
```diff
@@ -617,7 +617,13 @@ protected TestSpecification usingSpec(String name) {
      * @return the test specification; never null
      */
     protected TestSpecification usingSpec(String name, String directory) {
-        return usingSpec(name, Paths.get(directory));
+        String version = System.getProperty("java.version");
+        String modulePath = "" ;
+        //Java 11 has issues with determine paths when invoking tests from the root directory, so this is a temp workaround.
```
Comment (gunnarmorling) [inline debezium-embedded/src/test/java/io/debezium/embedded/ConnectorOutputTest.java]: So does this also work when running the build just from within this module?

File: pom.xml [hunk 6]
```diff
@@ -57,19 +57,19 @@
...
-        <version.confluent.platform>5.1.2</version.confluent.platform>
+        <version.confluent.platform>5.3.1</version.confluent.platform>
 
         <!-- Databases -->
         <version.postgresql.driver>42.2.8</version.postgresql.driver>
         <version.mysql.server>5.7</version.mysql.server>
-        <version.mysql.driver>8.0.16</version.mysql.driver>
-        <version.mysql.binlog>0.19.1</version.mysql.binlog>
+        <version.mysql.driver>8.0.17</version.mysql.driver>
+        <version.mysql.binlog>0.20.1</version.mysql.binlog>
```
Comment (gunnarmorling) [inline pom.xml]: Those dependency updates are good, but seem unrelated to JDK 11 migration; could you file a separate issue and change these in a separate commit attributed to that? Helps to highlight the change in the release notes.

Comment (brbrown25) [reply pom.xml]: Yep I can do that!

- [ ] Remove these updates, file separate jira.

Comment (brbrown25) [reply pom.xml]: Yeah I need to add a step to Travis to separately run tests under both jdks and tweak to still only build jdk8 byte code.

**Comment by brbrown25** [comment]
> @brbrown25 A few comments inline.
> 
> Also it seems now when _building_ with JDK 11, this will also create JDK 11 byte code? As discussed before, we should keep 1.8 compatibility at runtime to not exclude too many users.
> 
> Thanks a lot again for your efforts on this!



> Hey @brbrown25, how are things? Seems this last test failure might be unrelated? Do you observe it locally too on your branch?

Yep it fails locally for me as well. This weekend I’ll get to finally wrap up those remaining items. Had a big release this week so got a little side tracked. Almost there though!

**Comment by brbrown25** [comment]
@gunnarmorling it does appear that these changes are causing the mysql build failures, I just did a clean checkout of master branch they worked fine locally for me. gonna play around more to see why that one test is blowing up.

**Comment by gunnarmorling** [comment]
Hey @brbrown25, howdy? Thought I'd check in on this one; where are we with it, and what's needed to wrap it up? Thanks a lot!

**Comment by brbrown25** [comment]
> Hey @brbrown25, howdy? Thought I'd check in on this one; where are we with it, and what's needed to wrap it up (in any case, a rebase, as there are some conflicts)? Thanks a lot!

@gunnarmorling The only thing outstanding on this is figuring out why the mysql tests are failing. I plan to do a rebase today and take another look at this!

**Comment by brbrown25** [comment]
I’ll be resolving the latest conflict and taking another crack at these failing tests this weekend!

**Comment by brbrown25** [comment]
@gunnarmorling sorry for the long delay, been a little crazy with some work projects. I think I may have a bead on what's happening with the failing test and actively exploring it today. I've removed the code formatting commit as I did not mean to push that when I did my last rebase

**Comment by gunnarmorling** [comment]
Ok, thanks a lot, @brbrown25! Just wanted to come here and ask about it and then I saw your message :) To summarize, is the following correct?

With this change:

* Debezium can be built with JDK 8 and 11
* We'd keep building on JDK 8 for the time being
* The built and published binaries can be run on JDK 8 and 11
* We'd update the container images to JDK 11

Did I get this right?

**Comment by brbrown25** [comment]
> Ok, thanks a lot, @brbrown25! Just wanted to come here and ask about it and then I saw your message :) To summarize, is the following correct?
> 
> With this change:
> 
> * Debezium can be built with JDK 8 and 11
> * We'd keep building on JDK 8 for the time being
> * The built and published binaries can be run on JDK 8 and 11
> * We'd update the container images to JDK 11
> 
> Did I get this right?

Yep that is the game plan!

**Comment by gunnarmorling** [comment]
Ok, thanks for confirming! Just let me know then when it's done from your side. Thanks!

**Comment by gunnarmorling** [comment]
Hey @brbrown25 just checking in on this one. Do you have a timeline for completing it? If you won't get to it any time soon, no problem either, just let us know and someone else can take over. It'll need a rebase for sure. Would be nice to have it off the plate :) Thanks!

**Comment by brbrown25** [comment]
> Hey @brbrown25 just checking in on this one. Do you have a timeline for completing it? If you won't get to it any time soon, no problem either, just let us know and someone else can take over. It'll need a rebase for sure. Would be nice to have it off the plate :) Thanks!
@gunnarmorling sorry for the delay, I'm going to rebase and get this puppy updated and take another wack at it today. I'll ping later today but unfortunately I'll need to pass this off if I can't solve it today :(

**Comment by brbrown25** [comment]
@gunnarmorling I have not been able to track the cause of the failing test. Unfortunately at this time I won’t be able to devote much time to this issue. :( sorry for dragging this one out, it’s definitely very close!

**Comment by jpechane** [comment]
@gunnarmorling ok, it seems to me that the problem is related to MySQL executing integration tests in two runs (normal and with replication) but sharing the same VM. It seems that in this case the JDBC driver is loaded once by one classloader and when the next run tries to load it then the driver is available but `DriverManager.isDriverAllowed` rejects it because it is loaded by another classloader.

**Comment by jpechane** [comment]
@brbrown25 Thanks for the hard work! Only small change was needed - completed inhttps://github.com/debezium/debezium/pull/1147

**Comment by brbrown25** [comment]
> @brbrown25 Thanks for the hard work! Only small change was needed - completed in #1147

oh nice catch! thanks for the opportunity!
